feat(njs): expose capture group variables

TPXP opened the pull request:
### Proposed changes

This PR exposes capture group variables like `$1`, `$2` to njs scripts in the `Request.variables` object, avoiding duplicate parsing to determine this value (on nginx side and on njs side).

### Checklist

Before creating a PR, run through this checklist and mark each as complete:
- [x] I have read the [`CONTRIBUTING`](CONTRIBUTING.md) document
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works
- [ ] If applicable, I have checked that any relevant tests pass after adding my changes

I'm not very familiar with the testing framework of this project so I couldn't test everything, here's my test scenario though (it makes sure we handle limits correctly and ensures previous variables are still accessible).
<details>
  <summary>Testing scenario</summary>
    
```javascript
// njs/test.js
function handler(r) {
  for (let i = 0; i < 10; i++) {
    console.log(`${i}: ${r.variables[i.toString()]}`);
  }
  console.log(`host: ${r.variables["http_host"]}`);
  r.return(200, `OK ${r.variables.http_host}`);
}

export default {handler}
```

```
# nginx.conf
daemon off;
worker_processes 1;

# Make sure JS log messages will appear on the console
error_log stderr info;

events {
  worker_connections 1024;
}

pid /tmp/nginx/run.pid;


http {
  access_log off;

  # Get nginx to start
  client_body_temp_path /tmp/nginx/body_tmp;
  proxy_temp_path /tmp/nginx/proxy_tmp;
  fastcgi_temp_path /tmp/nginx/fcgi_tmp;
  uwsgi_temp_path /tmp/nginx/uwsgi;
  scgi_temp_path /tmp/nginx/scgi;

  js_path "./njs/";
  js_import test from test.js;
  default_type  text/plain;

  server {
    listen       9090;
    # Define various locations with more or less matching groups
    location = /no_match {
      js_content test.handler;
    }
    location ~ ^/(.+)/(.+)/(.+)/(.+) {
      js_content test.handler;
    }
    location ~ ^/(.+)/(.+)/(.+) {
      js_content test.handler;
    }
    location ~ ^/(.+)/(.+) {
      js_content test.handler;
    }
    location ~ ^/(.+) {
      js_content test.handler;
    }
  }
}
```

Get URLS:
* /no_match
* /abc/def/ghi/jkl
* /abc/def/ghi
* /abc/def
* /abc

Logs look like this
```
2024/07/30 18:00:34 [info] 53869#0: *1 js: 0: /abc/def/ghi/jkl
2024/07/30 18:00:34 [info] 53869#0: *1 js: 1: abc
2024/07/30 18:00:34 [info] 53869#0: *1 js: 2: def
2024/07/30 18:00:34 [info] 53869#0: *1 js: 3: ghi
2024/07/30 18:00:34 [info] 53869#0: *1 js: 4: jkl
2024/07/30 18:00:34 [info] 53869#0: *1 js: 5: undefined
2024/07/30 18:00:34 [info] 53869#0: *1 js: 6: undefined
2024/07/30 18:00:34 [info] 53869#0: *1 js: 7: undefined
2024/07/30 18:00:34 [info] 53869#0: *1 js: 8: undefined
2024/07/30 18:00:34 [info] 53869#0: *1 js: 9: undefined
2024/07/30 18:00:34 [info] 53869#0: *1 js: host: localhost:9090
```
</details>

---
Inspired by https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_variable.c#L42

xeioex commented on nginx/ngx_http_js_module.c:
```diff
@@ -2972,6 +2972,38 @@ ngx_http_js_request_variables(njs_vm_t *vm, njs_object_prop_t *prop,
 
     if (setval == NULL) {
+        #if (NGX_PCRE)
+        is_capture = 1;
```
Nitpicking: is_capture = 0; as a starting point seems more appropriate.

xeioex replied:
>The current logic is shorter, I think we should keep it like that.

Agree, disregard the comment.

xeioex commented on nginx/ngx_http_js_module.c:
```diff
@@ -2972,6 +2972,38 @@ ngx_http_js_request_variables(njs_vm_t *vm, njs_object_prop_t *prop,
+        }
+
+        if (is_capture) {
+            /*
```
The commit seems excessive here, because we have similar places with no such comments. It is a basic PCRE API knowledge.

xeioex reviewed (changes requested):
We also need autotests here. `nginx/t/js_capture_variables.t` can be added.
It can be based on `nginx/t/js_variables.t`. See `nginx/t/README` for more details.

Version check should be added as well:

```
TODO: {                                                                          
local $TODO = 'not yet' unless has_version('0.8.6');

<capture tests here>

}
```

xeioex reviewed (changes requested):
Proposed commit log:

`HTTP: expose capture group variables.`

TPXP reviewed (commented):
I've added a test, let me know if you have other remarks

xeioex commented:
Hi @TPXP,

I sent my minor improvements to your branch. Review them and squash the patches into a single one.
Otherwise looks good to me.

TPXP commented:
Thanks for the review, passing the index from the test simplifies the syntax - good idea. 👍 
I've just shortened the index fetching logic (to avoid converting the value to a number before converting it back to a string). https://github.com/nginx/njs/pull/769/files#diff-a5b346a0c232ffaa460fb68331af6fa52d12138b101fbcf5ec02bbea0fe0bca8R57

I've rebased and squashed commits, let me know if something else is missing.

xeioex commented on nginx/ngx_http_js_module.c:
```diff
@@ -2972,6 +2972,33 @@ ngx_http_js_request_variables(njs_vm_t *vm, njs_object_prop_t *prop,
     name.len = val.length;
 
     if (setval == NULL) {
+        #if (NGX_PCRE)
```
I just realized we do not need this #ifdef because njs module will not compile without nginx PCRE anyway.

TPXP replied:
Ah, I haven't checked that. Removed the check then 👍

TPXP commented:
By the way, I haven't received feedback yet for a companion nginx patch preventing duplicate calls to the js handler. Nothing urgent, but just to be sure could you confirm I've sent my patch at the right place to get it merged into nginx's core ? :)

https://mailman.nginx.org/pipermail/nginx-devel/2024-August/N7VFIYUKSZFUIAO24OJODKQGTP63R5NV.html

xeioex commented:
@TPXP 

> I've sent my patch at the right place to get it merged into nginx's core ? :)

Yes the place is correct. The team received the patch and will give a reply as time permits. Unfortunately it may take a while because they are busy with more priority work.

xeioex commented:
Thank you for the patch, committed.
